# Data Analytics Program
## Python Fundamentals for Data Analysts
## 4.6: Combining & Exporting data
## Alphy Jose

# Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Importing datasets

In [2]:
# Creating string of project folder path
path = r'C:\Users\augus\Documents\01-2026 Instacart Basket Analysis'

In [3]:
#Importing  the orders_checked Dataset
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index_col = False)

In [4]:
#Importing  the orders_products_prior Dataset
df_ords_prior= pd.read_csv(os.path.join(path,'02 Data','Original Data', 'orders_products_prior.csv'), index_col=False)

In [12]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [8]:
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [13]:
df_ords.shape

(3421083, 6)

In [14]:
df_ords_prior.shape

(32434489, 4)

# Step 1: Merge your prepared Instacart orders data with the new orders_products_prior dataframe 

In [12]:
# merging both dataframes
df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

#### Both dataframes have one common column which is 'order_id'. I didnt specify the type of join because the default typr for join is inner join. I am assuming i will get  full match from both dataframes. 

In [18]:
df_merged_large.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     NaN         196                  1          0   both  
1                     NaN       14084                  2          0   both  
2                     NaN       12427                  3          0   both  
3                     NaN       26088                  4          0   both  
4                     NaN       26405                  5          0   both

In [20]:
df_merged_large.shape

(32434489, 10)

#### The merged dataset have 32434489 rows which is same as the number of rows in df_ords_prior dataframe(largest dataframe).

In [19]:
#To sum up all values in the merge column
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

#### in the “_merge” column, we can see that there are only entries that have a value of “both"
#### I am checking how the merging is done using 2 random order id's


In [14]:
# printing rows where order_id is 2539329  from df_ords
df_ords[df_ords['order_id'] == 2539329]

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   

   days_since_prior_order  
0                     NaN

In [15]:
# # printing rows where order_id is 2539329  from df_ords_prior
df_ords_prior[df_ords_prior['order_id'] == 2539329]

order_id  product_id  add_to_cart_order  reordered
24076664   2539329         196                  1          0
24076665   2539329       14084                  2          0
24076666   2539329       12427                  3          0
24076667   2539329       26088                  4          0
24076668   2539329       26405                  5          0

In [16]:
# printing rows where order_id is 2539329  from df_merged_large
df_merged_large[df_merged_large['order_id'] == 2539329]

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     NaN         196                  1          0   both  
1                     NaN       14084                  2          0   both  
2                     NaN       12427                  3          0   both  
3                     NaN       26088                  4          0   both  
4                     NaN       26405                  5          0   both

In [17]:
# printing rows where order_id is 1  from df_ords
df_ords[df_ords['order_id'] == 1]

order_id  user_id  order_number  orders_day_of_week  \
1868044         1   112108             4                   4   

         order_hour_of_day  days_since_prior_order  
1868044                 10                     9.0

In [20]:
# printing rows where order_id is 1  from df_ords_prior
df_ords_prior[df_ords_prior['order_id'] == 1]

Empty DataFrame
Columns: [order_id, product_id, add_to_cart_order, reordered]
Index: []

In [19]:
# printing rows where order_id is 1  from df_merged_large
df_merged_large[df_merged_large['order_id'] == 1]

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, product_id, add_to_cart_order, reordered, _merge]
Index: []

#### For oder id 1 there are no records in df_ords_prior and df_merged_large dataframes.
#### For order id 2539329, there is only 1 record in df_ords, 5 records in df_ords_prior and 5 records in df_merged_large 
#### So the  column, “order_id,” doesnt exists completely in both dataframes
#### It is not actually a full match.
#### What pandas does here is fill in information about each product for every “order_id” in the df_ords dataframe, which is why the resulting dataframe has 32,434,489 rows (the same total count as the df_ords_prior dataframe). 
#### I am checking actual merge rate with outer join

In [6]:
# merging both dataframes with outer join
df_merged_test = df_ords.merge(df_ords_prior, on = 'order_id',how = 'outer' ,indicator = True)

In [7]:
df_merged_test['_merge'].value_counts()

_merge
both          32434489
left_only       206209
right_only           0
Name: count, dtype: int64

#### in the “_merge” column, we can see that there are 206209 entries that have a value of “left_only"

In [8]:
df_merged_test[df_merged_test['_merge'] == 'left_only']

order_id  user_id  order_number  orders_day_of_week  \
0                1   112108             4                   4   
145             17    36855             5                   6   
322             34    35220            20                   3   
328             36    79431            23                   6   
332             38    42756             6                   6   
...            ...      ...           ...                 ...   
32640490   3421054   153585             9                   5   
32640510   3421056    83898            13                   2   
32640516   3421058   136952            20                   3   
32640569   3421063   169679            30                   0   
32640611   3421070   139822            15                   6   

          order_hour_of_day  days_since_prior_order  product_id  \
0                        10                     9.0         NaN   
145                      15                     1.0         NaN   
322                      11                     8.0         NaN   
328                      18                    30.0         NaN   
332                      16                    24.0         NaN   
...                     ...                     ...         ...   
32640490                 12                     7.0         NaN   
32640510                 20                    11.0         NaN   
32640516                 18                    15.0         NaN   
32640569                 10                     4.0         NaN   
32640611                 10                     8.0         NaN   

          add_to_cart_order  reordered     _merge  
0                       NaN        NaN  left_only  
145                     NaN        NaN  left_only  
322                     NaN        NaN  left_only  
328                     NaN        NaN  left_only  
332                     NaN        NaN  left_only  
...                     ...        ...        ...  
32640490                NaN        NaN  left_only  
32640510                NaN        NaN  left_only  
32640516                NaN        NaN  left_only  
32640569                NaN        NaN  left_only  
32640611                NaN        NaN  left_only  

[206209 rows x 10 columns]

#### There are 206209 orders without  products in the df_ords_prior data set.
#### For this project I am proceeding with same inner join merge since we are working with datasets with full merge rate

# Step 2:Export the merged file in pickle format as “orders_products_combined.pkl”.

In [21]:
#export data to pickle
df_merged_large.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))